### ISSUE TITLE
Mutating job arguments results in unbounded key growth

### ISSUE THREAD

**Issue by nathanl** [lantins/resque-retry#155]
## TL;DR

If a job mutates its argument, an extraneous retry key is created in Redis and is never cleaned up. This results in unbounded memory growth in Redis. See [this demo](https://github.com/nathanl/bug-demo-resque-retry).

## Details

I'm doing some work for a client who noticed in their Redis instance that the number of `resque:resque-retry*` keys for their application grew continually. The keys were being created but never cleaned up.

I spent a lot of time investigating. The puzzling thing was that no exception was being raised in the job; even if I put a `rescue Exception` (generally a bad idea) inside of `def self.perform(data)`, nothing was rescued.

Eventually by a process of elimination, I realized that the line which caused the bug was one which mutated the argument hash given to the job. This was being done because the job passed its argument to a third-party library, which requires the hash it's given to contain a `"timestamp"` key with an actual `Time` object as the value; therefore, within the job, the code parsed the `"timestamp"` value to an object and updated the hash with it before calling the library method.

When the argument hash was mutated, it created an extra `resque:resque-retry` key. Eg, `keys resque:resque-retry*` might find:

```
1) "resque:resque-retry:MyJob:313f220a2bc0456affb755d501e11a15e9776412"
2) "resque:resque-retry:MyJob:03b9ec478a3f9bd5d97ba61b960e503bf345466e"
```

In the case of a genuine failure and retry, the key representing one of the variations of arguments would eventually be cleaned up. But the "extra" key would never be deleted.

A workaround for this issue would be simply "never mutate resque job arguments" - if you must do so, `.dup` them first.

I've created [a demo](https://github.com/nathanl/bug-demo-resque-retry) of this issue by forking this project and modifying the code in `examples/demo`. Follow the instructions in the README there to reproduce.

**Comment by nathanl** [comment]
This is a possible cause of https://github.com/lantins/resque-retry/issues/111 and https://github.com/lantins/resque-retry/issues/10

**Comment by jzaleski** [comment]
@nathanl it sounds like you have idea of how to fix this. Want to take a stab at it? I'm happy to help review any PRs you submit.

Thanks for the write-up!

**Comment by nathanl** [comment]
@jzaleski I spent some more time looking into this, and I think it's on `resque` itself to fix. `Resque::Failure.create` initializes a `Resque::Failure::MultipleWithRetrySuppression` and gives it the klass + arguments `payload` hash. By that point, they arguments may have already been mutated by the failed job. I think in case of failure, `resque` should repeat for the `Failure.create` call what it did when it initialized the original job: decode the raw JSON value it fetched from Redis, as this is guaranteed not to have been mutated, rather than assuming that the decoded `payload` is still as it was when first decoded.

I think that's all the time I have to put toward this issue, but at least there's a workaround (don't mutate args within your job!), and hopefully these are good breadcrumbs for anyone who wants to try fixing it.

**Comment by jzaleski** [comment]
Ok. Thank you for digging in further. If that's the case I agree that it would make more sense to fix it upstream. I'd prefer that approach vs. monkey-patching

**Comment by nathanl** [comment]
@jzaleski Maybe this is on `resque-retry` after all? See https://github.com/resque/resque/issues/1671#issuecomment-462465848

**Comment by clintmod** [comment]
As a workaround I did a deep clone on the job args before passing them to my jobs.

`cloned_args = Marshal.load(Marshal.dump(args))`

**Comment by jzaleski** [comment]
Great. If anyone is willing to submit a PR to address this issue
internally, I'd be happy to review / merge and cut a new release.

**Comment by davidcpell** [comment]
@nathanl @jzaleski 

I'm having a similar issue where I was getting started migrating resque data from one instance to another, and I noticed that one of our instances has 2.6 million Redis keys with barely anything showing in the resque web console. Almost all of the 2.6 million keys are resque:resque-retry:MyJob:abcdefg12345.

Any idea why the retry/delay tabs in resque web console are empty, and whether it's safe to remove these keys??

Also FWIW, I am seeing jobs with leftover :resque-retry: keys that do not mutate job arguments.

**Comment by davidcpell** [comment]
Also @nathanl I ran the demo and am either missing something or my experience my indicate the "mutated arguments" hypothesis is partially or wholly incorrect. In the demo I clicked "Create Failing Job with Retry" 5 times and only ended up with 1 leftover `resque:resque-retry:*` key.

**Comment by nathanl** [comment]
@davidcpell Weird. I don't think I'm going to investigate this further at this point; I found a solution ("dup before mutate") for the client I was working with and have moved on to projects in a different programming language. Hopefully this thread can at least be a good place to discuss the issue.

**Comment by clintmod** [comment]
@davidcpell they are safe to delete here's a script I used to clean up the keys older than 30 days: https://gist.github.com/clintmod/2d22e749e43c42bb234474010b14a6e9

**Comment by clintmod** [comment]
I also wrote a go program to do the same thing because I thought it would be faster. It didn't seem like it was. It took 11 hours to clean up 44 million keys. https://github.com/clintmod/delete-redis-keys/blob/master/main.go

**Comment by davidcpell** [comment]
thanks @clintmod! I was not aware of the `object` command - that's nifty!

**Comment by jzaleski** [comment]
@nathanl I'm glad you all found a work-around. Can you provide a bit more context on where you're `dup[ed]` the input(s) so that I can see about incorporating it into `resque-retry`?

**Comment by jzaleski** [comment]
In addition to the above configuration option, I made a few tweaks to `resque-retry` in commit: https://github.com/lantins/resque-retry/commit/2892a3f46077b04c4f9efddf81022e840e1dcb26 -- particularly in how it handles "retry args."

`resque-retry@1.7.0`, which has the aforementioned changes, has been released to _RubyGems_ (https://rubygems.org/gems/resque-retry/versions/1.7.0) and is available now. Feedback welcome!

**Comment by jzaleski** [comment]
Closing as a clear work-around was provided and some supporting code-changes have been committed and made available.

**Comment by nathanl** [comment]
> Can you provide a bit more context on where you're dup[ed] the input(s)

I did that within the job itself, just before it did its mutation. I used `dup` as you do in your diff, since I knew it was sufficient for the use case, but `dump/load` as @clintmod showed would be a more foolproof "deep dup" for the library to perform, not knowing whether users may (eg) mutate the actual values in a hash.

**Comment by jzaleski** [comment]
I'd suggest you look at the commit that is linked in the issue (above). It
might be completely unrelated, but I saw an opportunity to make things more
consistent, so I did.

---

If you think this was / is not sufficient or would potentially cause other
collateral damage I'd like to better understand this.

As mentioned previously, configuring a TTL seems like a better approach vs.
anything else documented above (e.g. mass deleting keys, etc.)
